torch/_export/passes/add_runtime_assertions_for_constraints_pass.py

```diff
@@ -37,41 +55,47 @@ def __init__(self) -> None:
         self.current_gm: Optional[torch.fx.GraphModule] = None
 
     def _process_shape_constraints(self, constraints) -> Dict[str, List[ConstraintSpec]]:
-        constraints_name_to_constraint: Dict[str, List[ConstraintSpec]] = defaultdict(
-            list
-        )
-
-        constraint_name_to_dim: Dict[str, Dict[int, List[Tuple[int, int]]]] = defaultdict(
-            lambda: defaultdict(list)
+        input_name_to_dim_constraints: Dict[str, ConstraintsContainer] = defaultdict(
+            lambda: ConstraintsContainer(defaultdict(list), defaultdict(list))
         )
-
-        for name in constraints:
-            for dim, min_val, max_val in constraints[name]:
-                min_max = (_convert_to_int(min_val), _convert_to_int(max_val))
-                constraint_name_to_dim[name][dim].append(min_max)
+        for name, shape_constraints in constraints.items():
+            for dim, min_val, max_val in shape_constraints.ranges:
+                input_name_to_dim_constraints[name].ranges[dim].append(
+                    (_convert_to_int(min_val), _convert_to_int(max_val))
+                )
+            for dim, other_name, other_dim in shape_constraints.equalities:
+                input_name_to_dim_constraints[name].equalities[dim].append(
+                    (other_name, other_dim)
+                )
 
         # Merge the constraints into a single list of constraints
-        for name, dim_constraints in constraint_name_to_dim.items():
-            for dim, constraints in dim_constraints.items():
-                min_vals = [x[0] for x in constraints]
-                max_vals = [x[1] for x in constraints]
-                min_val = sorted(min_vals, reverse=True)[0]
-                max_val = sorted(max_vals, reverse=False)[0]
+        input_name_to_constraints: Dict[str, List[ConstraintSpec]] = defaultdict(list)
+        for name, dim_constraints in input_name_to_dim_constraints.items():
+            for dim, range_constraints in dim_constraints.ranges.items():
+                if range_constraints:
+                    min_vals, max_vals = zip(*range_constraints)
+                    min_val = max(min_vals)
+                    max_val = min(max_vals)
+                    assert min_val <= max_val
+                    input_name_to_constraints[name].append(
+                        RangeConstraintSpec(dim=dim, min_val=min_val, max_val=max_val)
+                    )
+            for dim, eq_constraints in dim_constraints.equalities.items():
+                for other_name, other_dim in eq_constraints:
+                    input_name_to_constraints[name].append(
+                        EqualityConstraintSpec(dim=dim, other_name=other_name, other_dim=other_dim)
+                    )
```

Comment:
Yeah, for sure. It should be possible to completely replace the current `input_constraints` and `input_name_to_example_inputs` with the set of data structures we were talking about with symbol / (name, dim) indexing.

I mention `input_name_to_example_inputs` too because the only remaining use of it in the add runtime assertion pass is for specialization constraints.

I'm assuming you will do it, but I could do it here if you want.